src/Microsoft.Net.Http.Headers/HeaderUtilities.cs

```diff
@@ -0,0 +1,247 @@
+                double qualityValue;
+                if (double.TryParse(qualityParameter.Value, NumberStyles.AllowDecimalPoint,
+                    NumberFormatInfo.InvariantInfo, out qualityValue))
+                {
+                    return qualityValue;
+                }
+            }
+            return null;
+        }
+
+        internal static void CheckValidToken(string value, string parameterName)
+        {
+            if (string.IsNullOrEmpty(value))
+            {
+                throw new ArgumentException("An empty string is not allowed.", parameterName);
+            }
+
+            if (HttpRuleParser.GetTokenLength(value, 0) != value.Length)
+            {
+                throw new FormatException(string.Format(CultureInfo.InvariantCulture, "Invalid token '{0}.", value));
+            }
+        }
+
+        internal static void CheckValidComment(string value, string parameterName)
+        {
+            if (string.IsNullOrEmpty(value))
+            {
+                throw new ArgumentException("An empty string is not allowed.", parameterName);
+            }
+
+            int length;
+            if ((HttpRuleParser.GetCommentLength(value, 0, out length) != HttpParseResult.Parsed) ||
+                (length != value.Length)) // no trailing spaces allowed
+            {
+                throw new FormatException(string.Format(CultureInfo.InvariantCulture, "Invalid comment '{0}'.", value));
+            }
+        }
+
+        internal static void CheckValidQuotedString(string value, string parameterName)
+        {
+            if (string.IsNullOrEmpty(value))
+            {
+                throw new ArgumentException("An empty string is not allowed.", parameterName);
+            }
+
+            int length;
+            if ((HttpRuleParser.GetQuotedStringLength(value, 0, out length) != HttpParseResult.Parsed) ||
+                (length != value.Length)) // no trailing spaces allowed
+            {
+                throw new FormatException(string.Format(CultureInfo.InvariantCulture, "Invalid quoted string '{0}'.", value));
+            }
+        }
+
+        internal static bool AreEqualCollections<T>(ICollection<T> x, ICollection<T> y)
+        {
+            return AreEqualCollections(x, y, null);
+        }
+
+        internal static bool AreEqualCollections<T>(ICollection<T> x, ICollection<T> y, IEqualityComparer<T> comparer)
+        {
+            if (x == null)
+            {
+                return (y == null) || (y.Count == 0);
+            }
+
+            if (y == null)
+            {
+                return (x.Count == 0);
+            }
+
+            if (x.Count != y.Count)
+            {
+                return false;
+            }
+
+            if (x.Count == 0)
+            {
+                return true;
+            }
+
+            // We have two unordered lists. So comparison is an O(n*m) operation which is expensive. Usually
```

Comment:
This is called from SomeHeaderType.Equals. I could imagine one of the parameters coming from the client where they would be limited by header lengths, but I would expect the other to come from the application.